test/AutoDiff/transposing_attr_type_checking.swift

```diff
@@ -0,0 +1,318 @@
+  return x
+}
+
+@transposing(missingSelfRequirement, wrt: 0)
+func missingSelfRequirementT<T: Differentiable>(x: T) -> T {
+  return x
+}*/
+
+func differentGenericConstraint<T: Differentiable & BinaryFloatingPoint>(x: T)
+-> T where T == T.TangentVector {
+  return x
+}
+
+// expected-error @+2 {{type 'T' does not conform to protocol 'BinaryFloatingPoint'}}
+// expected-error @+1 {{could not find function 'differentGenericConstraint' with expected type '<T where T : Differentiable, T == T.TangentVector> (T) -> T'}}
+@transposing(differentGenericConstraint, wrt: 0)
+func differentGenericConstraintT<T: Differentiable>(x: T) 
+-> T where T == T.TangentVector {
+  return x
+}
+
+func transposingInt(x: Float, y: Int) -> Float {
+  if y >= 0 {
+    return x
+  } else {
+    return x
+  }
+}
+
+// expected-error @+1 {{can only transpose with respect to parameters that conform to 'Differentiable' and where 'Int == Int.TangentVector'}}
+@transposing(transposingInt, wrt: 1) 
+func transposingIntT1(x: Float, t: Float) -> Int {
+  return Int(x)
+}
+
+
+// expected-error @+1 {{'@transposing' attribute requires original function result to conform to 'Differentiable'}}
+@transposing(transposingInt, wrt: 0)
+func tangentNotLast(t: Float, y: Int) -> Float {
+  return t
+}
+
+// ~~~~~~~~~~~~~ Test methods. ~~~~~~~~~~~~~
+
+// // Method no parameters.
+extension Float {
+  func getDouble() -> Double {
+      return Double(self)
+  }
+}
+
+extension Double {
+  @transposing(Float.getDouble, wrt: self)
+  func structTranspose() -> Float {
+    return Float(self)
+  }
+}
+
+// Method with one parameter.
+extension Float {
+  func adding(_ double: Double) -> Float {
+    return self + Float(double)
+  }
+
+  @transposing(Float.adding, wrt: 0)
+  func addingT1(t: Float) -> Double {
+    return Double(self + t)
+  }
+
+  @transposing(Float.adding, wrt: self)
+  func addingT2(_ double: Double) -> Float {
+    return self + Float(double)
+  }
+    
+  @transposing(Float.adding, wrt: (self, 0))
+  func addingT3() -> (Float, Double) {
+    return (self, Double(self))
+  }
+}
+
+// Different self type/result type.
+extension Int {
+  func myAdding(_ double: Double) -> Float {
+    return Float(double)
+  }
+}
+
+extension Float {
+  @transposing(Int.myAdding, wrt: 0)
+  func addingT3(t: Int) -> Double {
+    return Double(self)
+  }
+
+  // TODO: throw an error due to Int not being differentiable.
+  // @transposing(Int.myAdding, wrt: (self, 0))
+  // func addingT3() -> (Int, Double) {
+  //   return (Int(self), Double(self))
+  // }
+}
+
+// Static method.
+struct A : Differentiable & AdditiveArithmetic {
+  public typealias TangentVector = A
+  var x: Double
+  
+  static prefix func -(a: A) -> A {
+    return A(x: -a.x)
+  }
+
+  @transposing(A.-, wrt: 0)
+  static func negationT(a: A) -> A {
+    return A(x: -a.x)
+  }
+}
+
+// Method with 3 parameters.
+extension Float {
+  func threeParams(_ x: Float, _ y: Double, _ z: Float) -> Double {
```

Comment:
There is a significant lack of test cases that use generics, which can be a source of bugs. Could you make some parameters use generics?

Reply:
Done added in https://github.com/apple/swift/pull/25684/commits/5e51b5ed912da6ed5fcbaf0e43e1c04ab1719902 ! Did find a bug because of it where I was adding the generic constraint always on the method, when I should be adding it on the self type if it's a method.